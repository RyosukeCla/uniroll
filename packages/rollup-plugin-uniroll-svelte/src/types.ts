export type ResolveIdFallback = (
  specifier: string,
  importer?: string
) => string | void;
