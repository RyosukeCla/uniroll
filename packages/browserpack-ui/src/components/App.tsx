import {
  Button,
  ButtonGroup,
  CSSReset,
  DarkMode,
  Flex,
  ThemeProvider
} from "@chakra-ui/core";
import React, { useCallback, useEffect, useState } from "react";
import { templateList } from "./constants";
import { FileEditorPane } from "./panes/EditorPanel/FileEditorPane";
import { FileSelectorPane } from "./panes/EditorPanel/FileSelectorPane";
import { RunnerPane } from "./panes/RunnerPane";
import { TemplatesPane } from "./panes/TemplatesPane";
import { VariablesPane } from "./panes/VariablesPane";
import { useEnv, IsInMobileContext } from "./contexts";
import { useWindowWidth } from "./hooks";
import { EditorPanel } from "./panes/EditorPanel";

export type Files = { [key: string]: string };
// export const Editor = React.lazy(() => import("./editor/Editor"));
export const extToLanguage: { [key: string]: string } = {
  ".ts": "typescript",
  ".tsx": "typescript",
  ".js": "javascript",
  ".css": "css"
};

export function App() {
  const width = useWindowWidth();
  return (
    <IsInMobileContext.Provider value={width < 768}>
      <DarkMode>
        <ThemeProvider>
          <CSSReset />
          <Internal />
        </ThemeProvider>
      </DarkMode>
    </IsInMobileContext.Provider>
  );
}
type SCENE = "editor" | "template" | "run" | "variables";
const firstFiles = templateList[0].files as any;
function Internal() {
  const env = useEnv();
  const [currentScene, setCurrentScene] = useState<SCENE>("editor");
  const [files, setFiles] = useState<Files>(firstFiles);
  const [currentFilepath, setCurrentFilepath] = useState<string | null>(null);
  useEffect(() => {
    (async () => {
      try {
        const data = await env.load();
        setFiles(data.files);
      } catch (err) {
        alert(err.message);
      }
    })();
  }, []);

  const onUpdate = useCallback(
    (filepath: string, value: string) => {
      const newFiles = { ...files, [filepath]: value };
      setFiles(newFiles);
      env.save({
        files: newFiles
      });
    },
    [files]
  );

  const onSelectFile = useCallback((filepath: string) => {
    setCurrentFilepath(filepath);
  }, []);

  return (
    <>
      <Flex direction="column" w="100%" h="100%">
        <Flex h="32px">
          <Header
            files={files}
            onClickTab={scene => {
              setCurrentScene(scene);
              setCurrentFilepath(null);
            }}
          />
        </Flex>
        <Flex h="calc(100% - 32px)">
          {currentScene == "run" && <RunnerPane files={files} />}
          {currentScene === "variables" && (
            <VariablesPane files={files} onUpdate={onUpdate} />
          )}

          {currentScene == "editor" && (
            <EditorPanel
              filepath={currentFilepath}
              files={files}
              onBack={() => setCurrentFilepath(null)}
              onUpdate={onUpdate}
              onSelectFile={onSelectFile}
            />
          )}
          {currentScene == "template" && (
            <TemplatesPane
              onSelectTemplate={async url => {
                const res = await fetch(url);
                const data = await res.json();
                setFiles(data.files);
                if (data.files["/variables.json"]) {
                  setCurrentFilepath(null);
                  setCurrentScene("variables");
                } else {
                  setCurrentScene("editor");
                }
              }}
            />
          )}
        </Flex>
      </Flex>
    </>
  );
}

function Header(props: { files: Files; onClickTab: (scene: SCENE) => void }) {
  return (
    <Flex w="100%">
      <Flex>
        <ButtonGroup pt={1} pl={1}>
          <Button size="sm" onClick={() => props.onClickTab("run")}>
            Run
          </Button>
          {props.files["/variables.json"] && (
            <Button size="sm" onClick={() => props.onClickTab("variables")}>
              Variables
            </Button>
          )}
          <Button size="sm" onClick={() => props.onClickTab("editor")}>
            Files
          </Button>
        </ButtonGroup>
      </Flex>
      <Flex flex={1} />
      <Flex pr={1}>
        <ButtonGroup pt={1} pl={1}>
          <Button size="sm" onClick={() => props.onClickTab("template")}>
            Templates
          </Button>
        </ButtonGroup>
      </Flex>
    </Flex>
  );
}
