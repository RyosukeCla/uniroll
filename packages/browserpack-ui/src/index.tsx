export { EnvContext, useEnv } from "./components/contexts";
export { App } from "./components/App";
