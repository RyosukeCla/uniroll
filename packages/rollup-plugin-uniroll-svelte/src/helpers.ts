import ts from "typescript";

// Example.
//     import foo from "foo";
// =>  import foo from "https://esm.sh/foo";
export const cdnRewriteTransformerFactory = (
  cdnPrefix: string | ((specifier: string) => string)
) => (ctx: ts.TransformationContext) => {
  function visitNode(node: ts.Node): ts.Node {
    if (ts.isImportDeclaration(node)) {
      const specifier = node.moduleSpecifier.getText();
      const trim = specifier.slice(1, specifier.length - 1);
      const result =
        typeof cdnPrefix === "string"
          ? rewriteSpecifier(trim, cdnPrefix)
          : cdnPrefix(trim);

      return ts.factory.updateImportDeclaration(
        node,
        node.decorators,
        node.modifiers,
        node.importClause,
        ts.factory.createStringLiteral(result)
      );
    }
    // return node;
    return ts.visitEachChild(node, visitNode, ctx);
  }

  return (source: ts.SourceFile) =>
    ts.factory.updateSourceFile(
      source,
      ts.visitNodes(source.statements, visitNode)
    );
};

export function extractImportSpecfiers(codes: string[]) {
  const urls: string[] = [];
  for (const code of codes) {
    const source = ts.createSourceFile(
      "file:///_.tsx",
      code,
      ts.ScriptTarget.ESNext
    );
    source.statements.forEach((node) => {
      if (ts.isImportDeclaration(node)) {
        const target = node.moduleSpecifier.getText(source);
        const trimed = target.substr(1, target.length - 2);
        if (urls.findIndex((t) => t === trimed) === -1) {
          urls.push(trimed);
        }
      }
    });
  }
  return urls;
}

const rewriteSpecifier = (specifier: string, cdnPrefix: string) => {
  // relative path
  if (specifier.startsWith(".")) {
    return specifier;
  }

  // https direct
  if (specifier.startsWith("https://")) {
    return specifier;
  }
  return `${cdnPrefix}${specifier}`;
};

export const defaultCompilerOptions: ts.CompilerOptions = {
  target: ts.ScriptTarget.ES2019,
  module: ts.ModuleKind.ESNext,
  moduleResolution: ts.ModuleResolutionKind.NodeJs,
  esModuleInterop: true,
  jsx: ts.JsxEmit.React,
  resolveJsonModule: true,
};
