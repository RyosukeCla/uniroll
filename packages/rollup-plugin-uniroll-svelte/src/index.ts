import { PreprocessorGroup } from "svelte/types/compiler/preprocess";
import type { Plugin } from "rollup";
import { compile as svelteCompile, preprocess } from "svelte/compiler";
import {
  cdnRewriteTransformerFactory,
  createSveltePreprocessor,
} from "./tsPreprocess";
import ts from "typescript";
import { CompileOptions } from "svelte/types/compiler/interfaces";
import { ResolveIdFallback } from "./types";

type SveltePluginOptions = {
  target: ts.ScriptTarget;
  resolveIdFallback?: ResolveIdFallback;
  svelteOptions?: CompileOptions;
  extraPreprocess?: PreprocessorGroup[];
};

const defaultResolveIdFallback: ResolveIdFallback = (
  id: string,
  importer?: string
) => {
  return;
};

export const svelte = (opts: SveltePluginOptions) => {
  /**
   * if css is set to false, then output css file instead of js injection.
   */
  const willOutptuCss = !opts.svelteOptions?.css;
  const cssFiles = new Map<string, any>();

  return {
    name: "uniroll-svelte",
    buildStart() {
      cssFiles.clear();
    },
    load(id: string) {
      if (!cssFiles.has(id)) return;
      return cssFiles.get(id);
    },
    async transform(code: string, id: string) {
      if (id.endsWith(".svelte")) {
        const tsPreprocess = createSveltePreprocessor({
          target: opts.target,
          importer: id,
          resolveIdFallback: opts.resolveIdFallback ?? defaultResolveIdFallback,
        });

        const { code: preprocessed } = await preprocess(
          code,
          [tsPreprocess, ...(opts.extraPreprocess ?? ([] as any))],
          {
            filename: id,
          }
        );
        const result = svelteCompile(preprocessed, opts.svelteOptions);
        if (result.warnings.length > 0) {
          this.warn(result.warnings.map((t) => t.message).join("\n"));
        }

        if (willOutptuCss && result.css.code) {
          const cssFileName = `/svelte-generated-css/${Math.random().toString(16)}.css`;
          result.js.code += `\nimport "${cssFileName}";\n`;
          cssFiles.set(cssFileName, result.css);
        }

        const ret = ts.transpileModule(result.js.code, {
          fileName: "$temp.tsx",
          compilerOptions: {
            target: opts.target,
            module: ts.ModuleKind.ESNext,
          },
          transformers: {
            before: [cdnRewriteTransformerFactory(opts.resolveIdFallback!, id)],
          },
        });

        return {
          code: ret.outputText,
          map: ret.sourceMapText,
        };
      }
    },
  } as Plugin;
};
