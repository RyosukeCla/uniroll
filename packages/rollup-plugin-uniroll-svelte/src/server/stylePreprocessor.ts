import {
  Preprocessor,
  PreprocessorGroup,
} from "svelte/types/compiler/preprocess";
// @ts-ignore
import precss from "precss";
import postcss from "postcss";
import autoprefixer from "autoprefixer";

export const createStylePreprocessor = (
  opts: {
    autoprefixer?: autoprefixer.Options;
  } = {}
): PreprocessorGroup => {
  const styleProcessor: Preprocessor = async ({
    content,
    attributes,
    filename,
  }) => {
    const result = await postcss([
      precss,
      autoprefixer(
        opts.autoprefixer ?? {
          // @ts-ignore
          grid: true,
          overrideBrowserslist: ["last 2 versions", "ie 11"],
        }
      ),
    ]).process("/* autoprefixer grid: autoplace */\n" + content, {
      from: undefined,
    });
    return {
      code: result.css,
      map: result.map,
    };
  };
  return {
    style: styleProcessor,
  } as PreprocessorGroup;
};
