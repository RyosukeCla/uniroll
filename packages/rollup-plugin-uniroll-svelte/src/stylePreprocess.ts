import { Preprocessor } from "svelte/types/compiler/preprocess";

export const createStylePreprocessor = ({}: {}) => {
  const style: Preprocessor = async ({ content, attributes, filename }) => {
    return {
      code: content,
    };
  };
  return {
    style,
  };
};

// @ts-ignore
// import precss from "precss";
// import postcss from "postcss";
// import autoprefixer from "autoprefixer";
// async function transformWithAutoprefixer(input: string) {
//   const result = await postcss([
//     precss,
//     autoprefixer({
//       // @ts-ignore
//       grid: true,
//       overrideBrowserslist: ["last 2 versions", "ie 11"],
//     }),
//   ]).process("/* autoprefixer grid: autoplace */\n" + input, {
//     from: undefined,
//   });
//   return result.css;
// }
