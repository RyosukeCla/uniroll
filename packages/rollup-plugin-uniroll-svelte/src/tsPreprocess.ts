import { Preprocessor } from "svelte/types/compiler/preprocess";
import ts from "typescript";
import { cdnRewriteTransformerFactory } from "uniroll";

export const createSveltePreprocessor = ({
  cdnPrefix = "https://esm.sh/",
  target,
}: {
  cdnPrefix?: string;
  target: ts.ScriptTarget;
}) => {
  const script: Preprocessor = async ({ content, attributes, filename }) => {
    const out = transpileSvelteTypeScript(content, {
      fileName: filename ?? "/$$.tsx",
      cdnPrefix: (id: string) => {
        if (id === "svelte" || id === "svelte/internal") {
          return id;
        }
        if (id.startsWith(".")) {
          return id;
        }
        return `${cdnPrefix}${id}`;
      },
      target,
    });
    return { code: out.outputText, map: out.sourceMapText };
  };
  return {
    script,
  };
};

const svelteTSCompilerOptions: ts.CompilerOptions = {
  moduleResolution: ts.ModuleResolutionKind.NodeJs,
  target: ts.ScriptTarget.ES5,
  module: ts.ModuleKind.ESNext,
  importsNotUsedAsValues: ts.ImportsNotUsedAsValues.Error,
  types: ["svelte"],
  isolatedModules: true,
  sourceMap: true,
  allowJs: true,
  strict: false,
  esModuleInterop: true,
  skipLibCheck: true,
  forceConsistentCasingInFileNames: true,
  resolveJsonModule: true,
};

function transpileSvelteTypeScript(
  code: string,
  opts: {
    fileName: string;
    cdnPrefix: (id: string) => string;
    target: ts.ScriptTarget;
  }
): ts.TranspileOutput {
  // 内部的に tsx 拡張子ということにする
  const out = ts.transpileModule(code, {
    fileName: opts.fileName ?? "/_.tsx",
    compilerOptions: { ...svelteTSCompilerOptions, target: opts.target },
    transformers: {
      before: [
        ...(opts.cdnPrefix
          ? [cdnRewriteTransformerFactory(opts.cdnPrefix)]
          : []),
        importTransformer,
      ],
    },
  });
  // debugger;
  return out;
}

// https://github.com/sveltejs/svelte-preprocess/blob/master/src/transformers/typescript.ts#L35-L54
const importTransformer: ts.TransformerFactory<ts.SourceFile> = (context) => {
  const visit: ts.Visitor = (node) => {
    if (ts.isImportDeclaration(node)) {
      if (node.importClause?.isTypeOnly) {
        return ts.factory.createEmptyStatement();
      }
      return ts.factory.createImportDeclaration(
        node.decorators,
        node.modifiers,
        node.importClause,
        node.moduleSpecifier
      );
    }
    return ts.visitEachChild(node, (child) => visit(child), context);
  };
  return (node) => ts.visitNode(node, visit);
};
