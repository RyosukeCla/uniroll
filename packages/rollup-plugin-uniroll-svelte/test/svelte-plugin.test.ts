import fetch from "isomorphic-unfetch";
import { bundle } from "uniroll";
import { svelte } from "../src";
import ts from "typescript";
import { Plugin } from "rollup";
import { createStylePreprocessor } from "../src/server/stylePreprocessor";
global.fetch = fetch;

it("bundle with svelte", async () => {
  const bundled = await bundle({
    cdnPrefix: "https://cdn.skypack.dev/",
    input: "/index.tsx",
    files: {
      "/index.tsx": `
import App from "./App.svelte";
new App({target: document.body});
      `,
      "/App.svelte": `
<script lang="ts">
  let x = 1;
</script>
<style>
  .text {
    color: red;
  }
</style>
<span class="text">{x}</span>
      `,
    },
    extraPlugins: [
      svelte({
        target: ts.ScriptTarget.ES2019,
        cdnPrefix: "https://esm.sh/",
        svelteOptions: {},
      }) as any,
    ],
  });
  const out = await bundled.generate({ format: "es" });
  expect(out.output[0].code).toMatchSnapshot();
});

it("bundle for es5", async () => {
  const bundled = await bundle({
    cdnPrefix: "https://cdn.skypack.dev/",
    input: "/index.tsx",
    files: {
      "/index.tsx": `
import App from "./App.svelte";
const foo = (x: string) => console.log(x);
foo();

new App({target: document.body});
      `,
      "/App.svelte": `
<script lang="ts">
  import { onMount } from "svelte";
  let x = 1;
  onMount(() => {
    console.log("xxxx");
  });
</script>
<style>
  .text {
    display: grid;
    color: red;
  }
</style>
<span class="text">{x}</span>
      `,
    },
    compilerOptions: {
      target: ts.ScriptTarget.ES5,
    },
    extraPlugins: [
      svelte({
        target: ts.ScriptTarget.ES5,
        cdnPrefix: "https://esm.sh/",
        extraPreprocess: [
          createStylePreprocessor({
            autoprefixer: {
              // @ts-ignore
              grid: true,
              overrideBrowserslist: ["last 2 versions", "ie 11"],
            },
          }),
        ],
        svelteOptions: {
          legacy: true,
        },
      }) as Plugin,
      {
        name: "transform-cdn",
        transform(code, id) {
          if (id.startsWith("https://cdn.skypack.dev/svelte")) {
            const out = ts.transpileModule(code, {
              compilerOptions: {
                module: ts.ModuleKind.ESNext,
                target: ts.ScriptTarget.ES5,
              },
            });
            return {
              code: out.outputText,
              map: out.sourceMapText,
            };
          }
        },
      } as Plugin,
    ],
  });
  const out = await bundled.generate({ format: "es" });
  expect(out.output[0].code).toContain("-ms-grid");
  expect(out.output[0].code).toContain("/** @class */");
});
