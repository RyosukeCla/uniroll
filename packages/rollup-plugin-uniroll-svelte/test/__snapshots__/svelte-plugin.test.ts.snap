// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle with svelte 1`] = `
"import { SvelteComponent, init, safe_not_equal, element, append, attr, insert, noop, detach } from 'svelte/internal';

/* generated by Svelte v3.29.7 */

function add_css() {
	var style = element(\\"style\\");
	style.id = \\"svelte-17agfn1-style\\";
	style.textContent = \\".text.svelte-17agfn1{color:red}\\";
	append(document.head, style);
}

function create_fragment(ctx) {
	let span;

	return {
		c() {
			span = element(\\"span\\");
			span.textContent = \`\${x}\`;
			attr(span, \\"class\\", \\"text svelte-17agfn1\\");
		},
		m(target, anchor) {
			insert(target, span, anchor);
		},
		p: noop,
		i: noop,
		o: noop,
		d(detaching) {
			if (detaching) detach(span);
		}
	};
}

let x = 1;

class Component extends SvelteComponent {
	constructor(options) {
		super();
		if (!document.getElementById(\\"svelte-17agfn1-style\\")) add_css();
		init(this, options, null, create_fragment, safe_not_equal, {});
	}
}

new Component({ target: document.body });
"
`;
